sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/crypto/SValueHash.scala

```diff
@@ -0,0 +1,255 @@
+        case SValue.SText(v) =>
+          addByte(TEXT_TAG, "TEXT_TAG")
+            .addString(v)
+        case SValue.SContractId(cid) =>
+          addByte(CONTRACT_ID_TAG, "CONTRACT_ID_TAG")
+            .addCid(cid)
+        case SValue.SOptional(opt) =>
+          addByte(OPTIONAL_TAG, "OPTIONAL_TAG")
+            .addOptional(opt)
+        case SValue.SList(xs) =>
+          addByte(LIST_TAG, "LIST_TAG")
+            .addList(xs)
+        case SValue.SRecord(id, fs, v) =>
+          addByte(RECORD_TAG, "RECORD_TAG")
+            .addRecord(id, fs, v)
+        case SValue.SVariant(id, variant, rank, v) =>
+          addByte(VARIANT_TAG, "VARIANT_TAG")
+            .addVariant(id, variant, rank, v)
+        case SValue.SEnum(id, ctor, rank) =>
+          addByte(ENUM_TAG, "ENUM_TAG")
+            .addEnum(id, ctor, rank)
+        case SValue.SMap(isTextMap, entries) =>
+          if (isTextMap) {
+            addByte(TEXT_MAP_TAG, "TEXT_MAP_TAG")
+              .addTextMap(entries)
+          } else {
+            addByte(GEN_MAP_TAG, "GEN_MAP_TAG")
+              .addGenMap(entries)
+          }
+        case _: SValue.SAny | _: SValue.SBigNumeric | _: SValue.SPAP | _: SValue.SStruct |
+            SValue.SToken | _: SValue.STypeRep =>
+          throw new IllegalArgumentException(s"Unexpected SValue during hashing: $svalue")
+      }
+
+    private def addCid(cid: Value.ContractId): this.type =
+      addBytes(cid2Bytes(cid), s"${cid.coid} (contractId)")
+
+    private def addList(vs: FrontStack[SValue]): this.type =
+      iterateOver(vs.toImmArray)(_ addSValue _)
+
+    private def addVariant(
+        id: Identifier,
+        variant: VariantConName,
+        constructorRank: Int,
+        value: SValue,
+    ): this.type =
+      addQualifiedName(id.qualifiedName)
+        .addString(variant)
+        .addInt(constructorRank)
+        .addSValue(value)
+
+    private def addEnum(
+        id: Identifier,
+        constructor: Name,
+        constructorRank: Int,
+    ): this.type =
+      addQualifiedName(id.qualifiedName)
+        .addString(constructor)
+        .addInt(constructorRank)
+
+    private def addOptional(opt: Option[SValue]): this.type =
+      opt match {
+        case Some(v) => addByte(1.toByte, "Some (optional)").addSValue(v)
+        case None => addByte(0.toByte, "None (optional)")
+      }
+
+    private def addRecord(
+        id: Identifier,
+        labels: ImmArray[Name],
+        values: ArraySeq[SValue],
+    ): this.type = {
+      discard(addQualifiedName(id.qualifiedName))
+      val trailingNonesSize = values.reverseIterator.takeWhile {
+        case SOptional(None) => true
+        case _ => false
+      }.size
+      labels.iterator.take(labels.length - trailingNonesSize).zip(values).foreach { case (l, v) =>
+        addString(l).addSValue(v)
+      }
+      // This delimits the end of the record. Note that this does not collide with any of the tags.
+      addByte(0xff.toByte, "record end")
```

Comment:
maybe we can create a constant for that and put it with the comment together with the tags. 
I think you should at least put a comment above the tags, saying they should be different form 0xff.